### PR TITLE
Don't vote for empty leader transmissions

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Currently there's a check in pohrecorder -> reset() that skips resetting to a specific hash, slot index if the PohRecorder has already generated that hash in its tick cache.

However, this causes a bug when a node votes on an empty leader transmission (a slot full of ticks):

Example: 

A validator votes at slot 1, and thus starts a PohRecorder for slot 1. Then they receive nothing but ticks for slot 2. Meanwhile the local PohRecorder also ticks past 2 with all ticks, so we vote on slot 2, but don't reset the PohRecorder due to that check in PohRecorder. Now if the validator is leader at slot 3, then our current code will set parent to 2, but PohRecorder will send all the ticks from slot 1 + 2 b/c it was never reset.

#### Summary of Changes

Don't vote for empty leader transmissions.  This and the change here: https://github.com/solana-labs/solana/pull/3238 should fix the problem.

Fixes #

### CODE REVIEW

**Comment by carllin** [comment]
@aeyakovenko, we moved the check for empty leader transmissions to the bank because it seemed like a more natural home. Thanks @rob-solana  for all the help!

**Comment by garious** [comment]
You've basically written a regression test in the PR description. Can you encode it?

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -871,6 +877,11 @@ impl Bank {
     pub fn get_epoch_and_slot_index(&self, slot: u64) -> (u64, u64) {
         self.epoch_schedule.get_epoch_and_slot_index(slot)
     }
+
+    pub fn is_votable(&self) -> bool {
+        let max_tick_height = (self.slot + 1) * self.ticks_per_slot - 1;
+        self.is_delta.load(Ordering::Relaxed) && self.tick_height() == max_tick_height
```
Comment (garious) [inline runtime/src/bank.rs]: This pub function (called from another crate) needs tests and a doc comment. At first glance, it looks like this implementation can be replaced with:

```
pub is_votable(&self) -> bool {
   self.is_frozen() && !self.accounts.is_empty()
}
```

but no way to find out from within the solana_runtime crate.  The test up in ReplayStage only tests one case, but this function alone has several edge cases.

Comment (carllin) [reply runtime/src/bank.rs]: Those are not equivalent, is_delta can be false when !self.accounts.is_empty() is true (f your parent bank is nonempty, the child bank will always be nonempty as well, even if the child processes no transactions). We check self.tick_height() == max_tick_height instead of is_frozen because @rob-solana pointed out we may want to check if a bank is votable before freezing it.

Comment (carllin) [reply runtime/src/bank.rs]: Also, which edge cases were you thinking of. The one I can think of is that a bank that processes only ticks has self.is_delta =- false

Comment (rob-solana) [reply runtime/src/bank.rs]: @carllin there definitely should be local tests.  cargo test --manifest-path runtime/Cargo.toml should cover everything in here

@garious this check is conservative and so shouldn't have edge cases.  once tick_height() is max, no commit() will be called, and commit() is where we move to "is_delta".

the reason this check is down here is because otherwise, its logic would have to be duplicated in the TVU and in the TPU (which talk to the bank at different levels)

Comment (garious) [reply runtime/src/bank.rs]: Why not use HashQueue::hash_height()? I don't really want an answer as much as I want a test to answer on your behalf.
